Allow direct field access in Java Forms

mkurz opened the pull request:
Fixes https://discuss.lightbend.com/t/databinder-field-access-instead-of-using-accessors/2468

Adds `play.forms.binding.directFieldAccess` config to switch on direct field access for all forms. To control specific forms only this pull request adds `form.withDirectFieldAccess(...)` to switch direct field access on or off - or if `null` falls back to config fallback.

marcospereira commented on framework/src/play-java-forms/src/main/java/play/data/DynamicForm.java:
```diff
@@ -226,6 +226,15 @@ public DynamicForm withLang(Lang lang) {
     }
 
+    @Override
+    public DynamicForm withDirectFieldAccess(Boolean directFieldAccess) {
```
Why not use a `boolean` instead of a `Boolean`?

mkurz replied:
I change that. Originally I choose a different approach and that `Boolean` was a left-over which now made this pr a mess. I fixed that now.

marcospereira commented on framework/src/play-java-forms/src/main/java/play/data/Form.java:
```diff
@@ -184,6 +185,26 @@ public Form(String rootName, Class<T> clazz, Map<String,String> data, List<Valid
+     * @param lang used for formatting when retrieving a field (via {@link #field(String)} or {@link #apply(String)}) and for translations in {@link #errorsAsJson()}
+     * @param directFieldAccess access fields of form directly during binding instead of using getters
+     */
+    public Form(String rootName, Class<T> clazz, Map<String,String> data, List<ValidationError> errors, Optional<T> value, Class<?>[] groups, MessagesApi messagesApi, Formatters formatters, ValidatorFactory validatorFactory, Config config, Lang lang, Boolean directFieldAccess) {
```
Same question about `boolean` vs `Boolean`.

marcospereira commented on framework/src/play-java-forms/src/main/java/play/data/Form.java:
```diff
@@ -444,6 +466,10 @@ private DataBinder dataBinder(String... allowedFields) {
         dataBinder.setValidator(validator);
         dataBinder.setConversionService(formatters.conversion);
         dataBinder.setAutoGrowNestedPaths(true);
+        if((this.directFieldAccess != null && this.directFieldAccess) || (this.directFieldAccess == null && this.config.getBoolean("play.forms.binding.directFieldAccess"))) {
```
I would use `boolean` default the value at the constructor to what is configured by `play.forms.binding.directFieldAccess`. This would make this check easier.

mkurz replied:
Done.

marcospereira commented on framework/src/play-java-forms/src/main/resources/reference.conf:
```diff
@@ -4,4 +4,10 @@ play {
+
+  forms {
+    binding {
+      directFieldAccess = false
```
Add some comment document this property.

mkurz replied:
Comment added.

marcospereira commented on framework/src/play-java-forms/src/test/scala/play/data/DynamicFormSpec.scala:
```diff
@@ -80,6 +80,13 @@ class DynamicFormSpec extends Specification {
       form("foo").value().get() must_== "bar"
     }
 
+    "fail with exception when trying to switch on direct field access" in {
```
Also, have a test for when passing `false`.

mkurz replied:
Test added.

mkurz commented on framework/src/play-java-forms/src/main/java/play/data/Form.java:
```diff
@@ -184,6 +185,26 @@ public Form(String rootName, Class<T> clazz, Map<String,String> data, List<Valid
      * @param lang used for formatting when retrieving a field (via {@link #field(String)} or {@link #apply(String)}) and for translations in {@link #errorsAsJson()}
      */
     public Form(String rootName, Class<T> clazz, Map<String,String> data, List<ValidationError> errors, Optional<T> value, Class<?>[] groups, MessagesApi messagesApi, Formatters formatters, ValidatorFactory validatorFactory, Config config, Lang lang) {
+        this(rootName, clazz, data, errors, value, groups, messagesApi, formatters, validatorFactory, config, lang, config != null && config.getBoolean("play.forms.binding.directFieldAccess"));
```
I checked for `config != null` because we have a test to not fail when all components are `null`:
https://github.com/playframework/playframework/blob/c562d81a7a5b946d29c3eee1be1557581d79bba7/framework/src/play-java-forms/src/test/scala/play/data/DynamicFormSpec.scala#L83-L86

mkurz commented:
@marcospereira I addressed all your comments plus added two more tests.

I would definitely **not** change `play.forms.binding.directFieldAccess` to `true` by default.

First of all you have to know it's either / or. Either you choose `initDirectFieldAccess` (which tries to access fields directly) or `initBeanPropertyAccess` (which tries to use getters and is the default). It's just not possible to mix these two.

These points come into my mind:
1) If we switch to `initDirectFieldAccess=true` then the getter methods would not be called anymore - but maybe in some cases such getters are not simple getters but execute logic before returning a value (even if it's only trimming a string before returning it). That logic wouldn't be executed anymore -> not backward compatible.

2) There may be security problems. Direct field access tries to make private fields accessible via reflection ([this](https://github.com/spring-projects/spring-framework/blob/v5.1.2.RELEASE/spring-beans/src/main/java/org/springframework/beans/DirectFieldAccessor.java#L128-L138) calls [that](https://github.com/spring-projects/spring-framework/blob/v5.1.2.RELEASE/spring-core/src/main/java/org/springframework/util/ReflectionUtils.java#L477)). That could cause security exceptions. We don't know which JVMs and/or security managers/settings Play users run their Play apps with, which may don't allow setting (certain) form fields accessible. Again not backward compatible.

3) I am not 100% sure how that direct field access works in regards to forms. Would it maybe try to access other private fields as well, which are not meant to be form fields and do something with them? I don't know.

As you can see even though I think that our tests would pass (because it looks in our code it's not a problem to make the private fields accessible and we use simple getters without logic everywhere) I would not change the default because that possible would break backward compatibility in user apps.

marcospereira commented:
Thanks for the explanation, @mkurz.

I agree that we should not change the default to `true` (I was under the impression that the binder was smart enough to handle both cases).

> There may be security problems. Direct field access tries to make private fields accessible via reflection (this calls that). That could cause security exceptions. We don't know which JVMs and/or security managers/settings Play users run their Play apps with, which may don't allow setting (certain) form fields accessible. Again not backward compatible.

Hum... I can foresee a warning about illegal reflective access in the future then when using Java 8+. But still, this is a configuration that is off by default. Anyway, it would be good to have a note/warning on our docs about it. Could you add a section on `JavaForms` that explains the use of `directFieldAccess` and its consequences?

mkurz commented:
@marcospereira I added docs that shows the usage of direct field access and mentions the risk of warnings and exceptions.
Please have a look once more, thanks Marcos!

mkurz commented:
@marcospereira Are the added docs ok for you?

marcospereira commented:
Backport to 2.7.x: a262d9029ca83bbdfd81b28f154822fe69dbf9cf
